lib/utils.js

```diff
@@ -146,35 +146,53 @@ function findToken (str, token) {
   return ind
 }
 
-const RDS1 = /^\.\.?\//u
-const RDS2 = /^\/\.(?:\/|$)/u
-const RDS3 = /^\/\.\.(?:\/|$)/u
-const RDS5 = /^\/?(?:.|\n)*?(?=\/|$)/u
-
-function removeDotSegments (input) {
+// RFC 5.2.4
+function removeDotSegments (path) {
+  let input = path
   const output = []
 
-  while (input.length) {
-    if (input.match(RDS1)) {
-      input = input.replace(RDS1, '')
-    } else if (input.match(RDS2)) {
-      input = input.replace(RDS2, '/')
-    } else if (input.match(RDS3)) {
-      input = input.replace(RDS3, '/')
-      output.pop()
+  while (input.length > 0) {
```

Comment:
I guess, it could be even optimized more by counting up positions we need to slice, and at the end we make a final slice.

e.g. '../../../../../index.html' will result in 5 slice-calls. But we could collect the amount of the characters, we could throw away, and just make one slice.

Maybe we should do it in a follow up?